### PR TITLE
Apply AD API deployment and serviceaccount assosication with ad detectors podtemplate

### PR DESCRIPTION
**Pull request by freecaykes**
## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/render/intrusion_detection.go
Comment (rene-dekker) [inline pkg/render/intrusion_detection.go:1188]: nit: throughout the render package we don't name the methods `getX`, but just `X`.

File: pkg/render/intrusion_detection.go
Comment (rene-dekker) [inline pkg/render/intrusion_detection.go:1307]: Could you please add a small comment in the code as to why we want this toleration?

File: pkg/render/intrusion_detection.go
Comment (rene-dekker) [inline pkg/render/intrusion_detection.go:58]: This field is used a s service account name. We don't typically use periods in service account names.

File: pkg/render/intrusion_detection.go
Comment (freecaykes) [reply pkg/render/intrusion_detection.go]: https://github.com/tigera/operator/pull/1931/commits/136b82e64ba9d34bee78baa945ee24909eb56e53 fixed here

File: pkg/render/intrusion_detection.go
Comment (freecaykes) [reply pkg/render/intrusion_detection.go:58]: https://github.com/tigera/operator/pull/1931/commits/136b82e64ba9d34bee78baa945ee24909eb56e53 fixed here

File: pkg/render/intrusion_detection.go
Comment (freecaykes) [reply pkg/render/intrusion_detection.go:1298]: thought this was a default config to add all our components in the control plane / master node, but now that I think about it would it be a good idea not to add this toleration since it needs disk space using `emptyDir` and might be better that the pod spawns on a worker node @doublek forgot if we chatted about this

Comment (freecaykes) [reply pkg/render/intrusion_detection.go]: chatted with @doublek deployment should respect control plane node but not master 

https://github.com/tigera/operator/pull/1931/commits/1d1cd5c744eda16cb0ba70b7220849402aab16d0 updated with comments

File: pkg/render/intrusion_detection.go
Comment (doublek) [inline pkg/render/intrusion_detection.go:1382]: I mentioned this just now in the other PR, this should be `detectors.tigera.io` and not what the `ADJobPodTemplateBaseName` refers to.

File: pkg/render/intrusion_detection.go
Comment (doublek) [inline pkg/render/intrusion_detection.go:161]: I think this file is getting large enough that I wonder if the AD objects should be refactored to a separate file. But I am okay to do that in a separate PR after getting this merged if the maintainers are okay with it for now.

File: pkg/render/intrusion_detection.go
Comment (doublek) [inline pkg/render/intrusion_detection.go:1329]: Not sure how there are handled in the other places of the operator. Is there any precedent for these numbers?

**Review by doublek — COMMENTED** [review]
I think overall it's looking good. I've added some initial comments.

File: pkg/render/intrusion_detection.go
Comment (doublek) [reply pkg/render/intrusion_detection.go]: I wonder if I provided the wrong guidance by suggestion using ControlPlaneSelectors? I thought the intention of these were to make sure Tigera components run on the right nodes. If I've misunderstood the intent I am okay to revert this.

File: pkg/render/intrusion_detection.go
Comment (freecaykes) [reply pkg/render/intrusion_detection.go]: going to remove them and use k8s default https://github.com/tigera/operator/pull/1931/commits/243589f6f789db15732483f8ba163a1b349e89bc

File: pkg/render/intrusion_detection.go
Comment (freecaykes) [reply pkg/render/intrusion_detection.go:1377]: https://github.com/tigera/operator/pull/1931/commits/243589f6f789db15732483f8ba163a1b349e89bc

**Comment by freecaykes** [comment]
build error waiting on https://github.com/tigera/operator/pull/1934/files - will merge this in as soon as it gets in

File: pkg/render/intrusion_detection.go
Comment (doublek) [inline pkg/render/intrusion_detection.go:1517]: typo: STORAGE

**Review by doublek — APPROVED** [review]
One typo. LGTM otherwise.

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and delete the branch after I've merged it.

File: pkg/controller/intrusiondetection/intrusiondetection_controller.go
Comment (rene-dekker) [inline pkg/controller/intrusiondetection/intrusiondetection_controller.go:438]: This should be:
```
rcertificatemanagement.NewKeyPairOption(intrusionDetectionCfg.ADAPIServerCertSecret, true, true),
```
The second param ensures that the secret is also created in the operator namespace. The controller responsible for creating the secret is the controller that should have this value set to true.

For context: A few lines down the value  is false for the typhaNodeTLS.NodeSecret. This is because the core_controller is the controller responsible for creating that secret.

File: pkg/render/intrusion_detection.go
Comment (rene-dekker) [inline pkg/render/intrusion_detection.go:1233]: "AD API Service and Deployment"
Is the comment maybe outdated? I think you can remove it, since it is redundant.

File: pkg/render/intrusion_detection.go
Comment (rene-dekker) [inline pkg/render/intrusion_detection.go:1264]: nit: I do not think there is a reason to put the envvars in a variable as opposed to just copying the lines into the &appsv1.Deployment{}

File: pkg/render/intrusion_detection.go
Comment (rene-dekker) [reply pkg/render/intrusion_detection.go:1233]: Update: I realize now that you have two functions related to creating service accounts. The comment makes sense now. No need to change it.

File: pkg/controller/intrusiondetection/intrusiondetection_controller.go
Comment (rene-dekker) [inline pkg/controller/intrusiondetection/intrusiondetection_controller.go:436]: You also need to add the service account `ADAPIObjectName` to it. This is the service account that will create a CertificateRequest when CertificateManagement is enabled. If the service account is missing here, it will not get the right RBAC for that operation.

**Comment by marvin-tigera** [comment]
Removing "merge-when-ready" label due to new commits

File: pkg/controller/intrusiondetection/intrusiondetection_controller.go
Comment (freecaykes) [reply pkg/controller/intrusiondetection/intrusiondetection_controller.go]: https://github.com/tigera/operator/pull/1931/commits/6ebf6fd366dd2d8e0d814b828b68b4a8af56d9a4

File: pkg/render/intrusion_detection.go
Comment (freecaykes) [reply pkg/render/intrusion_detection.go]: https://github.com/tigera/operator/pull/1931/commits/4546b11522d4ea1557bf00850577c3a07a6522aa

File: pkg/controller/intrusiondetection/intrusiondetection_controller.go
Comment (freecaykes) [reply pkg/controller/intrusiondetection/intrusiondetection_controller.go]: https://github.com/tigera/operator/pull/1931/commits/4546b11522d4ea1557bf00850577c3a07a6522aa

**Review by rene-dekker — APPROVED** [review]
lgtm

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and delete the branch after I've merged it.

**Comment by marvin-tigera** [comment]
I'm sorry but I failed delete the branch after merging the pull request.

File: pkg/render/intrusion_detection.go
Comment (tmjd) [inline pkg/render/intrusion_detection.go:121]: Both this and the adAPIImage are only needed when `!c.cfg.ManagedCluster` like the `jobInstallerImage` as far as I can tell. Is that correct?

**Pull request merged**
